src/utils/expansion.rs

```diff
@@ -13,8 +14,79 @@ use crate::RoverResult;
 
 /// Implements router-config-style
 /// [variable expansion](https://www.apollographql.com/docs/router/configuration/overview/#variable-expansion)
-/// for use in Rover configs (like `supergraph.yaml`).
-pub(crate) fn expand(value: &str) -> RoverResult<String> {
+/// for a YAML mapping (e.g., an entire `supergraph.yaml` or `router.yaml`).
+pub(crate) fn expand(value: Value) -> RoverResult<Value> {
+    match value {
+        Value::String(s) => expand_str(&s).map(Value::String),
+        Value::Null | Value::Bool(_) | Value::Number(_) | Value::Tagged(_) => Ok(value),
+        Value::Sequence(inner) => inner
+            .into_iter()
+            .map(expand)
+            .collect::<RoverResult<Sequence>>()
+            .map(Value::Sequence),
+        Value::Mapping(inner) => inner
+            .into_iter()
+            .map(|(key, value)| expand(value).map(|value| (key, value)))
+            .collect::<RoverResult<Mapping>>()
+            .map(Value::Mapping),
+    }
+}
+
+#[cfg(test)]
+mod test_expand {
+    use serde_yaml::Value;
+
+    #[test]
+    fn mapping() {
+        let yaml = "supergraph:\n  introspection: true\n  listen: 0.0.0.0:${env.PORT:-4000}";
+        let value = serde_yaml::from_str(yaml).unwrap();
+        let expanded = super::expand(value).unwrap();
+        let expected: Value =
+            serde_yaml::from_str("supergraph:\n  introspection: true\n  listen: 0.0.0.0:4000")
+                .unwrap();
+        assert_eq!(expanded, expected);
+    }
+
+    /// A realish world test of complex expansion
+    #[test]
+    fn router_config_header_injection() {
+        let yaml = r#"headers:
+  subgraphs:
+    products:
+      request:
+        - insert:
+            name: Router-Authorization
+            value: ${env.ROUTER_AUTHORIZATION:-test}
+    reviews:
+      request:
+        - insert:
+            name: Router-Authorization
+            value: ${env.ROUTER_AUTHORIZATION:-test2}"#;
+        let value = serde_yaml::from_str(yaml).unwrap();
+        let expanded = super::expand(value).unwrap();
+        let expected: Value = serde_yaml::from_str(
+            r#"headers:
```

Comment:
can we use `supergraph.yaml` examples here instead of `router.yaml` examples?

Reply:
Done!